### PR TITLE
Engine support for arbitrary shape physical layers

### PR DESCRIPTION
**Pull request by amirh**
Note that we just use the bounding rectangle on Fuchsia as it currently does not support arbitrary shape physical layers.

### CODE REVIEW

File: flow/layers/physical_model_layer.h [hunk 1]
```diff
@@ -35,12 +39,83 @@ class PhysicalModelLayer : public ContainerLayer {
...
+  const SkRect& getBounds() const { return rrect_.getBounds(); }
+
+  // |flow::PhysicalLayerShape|
+  const SkPath getPath() const;
+
+  // |flow::PhysicalLayerShape|
+  void clipCanvas(SkCanvas& canvas) const { canvas.clipRRect(rrect_, true); }
+
+  // |flow::PhysicalLayerShape|
+  bool isRect() const { return rrect_.isRect(); }
```
Comment (Hixie) [inline flow/layers/physical_model_layer.h]: I would rename this to `useLayer` or `compositeWithLayer` or some such.

**Comment by Hixie** [comment]
Logic LGTM, but I recommend getting someone more familiar with the engine's conventions to do a review of the C++-specific aspects.

**Comment by jason-simmons** [comment]
LGTM

File: flow/layers/physical_model_layer.h [hunk 2]
```diff
@@ -35,12 +39,83 @@ class PhysicalModelLayer : public ContainerLayer {
...
+  virtual const SkRRect& getFrameRRect() const = 0;
+#endif  // defined(OS_FUCHSIA)
+};
+
+class PhysicalLayerRRect : public PhysicalLayerShape {
+ public:
+  PhysicalLayerRRect(const SkRRect& rrect) { rrect_ = rrect; }
+
+  // |flow::PhysicalLayerShape|
+  const SkRect& getBounds() const { return rrect_.getBounds(); }
```
Comment (chinmaygarde) [inline flow/layers/physical_model_layer.h]: You are missing the `override` declaration on these. I will try to add a `clang-tidy` check for this. Thanks for decorating the method with a comment state what this overrides though.

File: flow/layers/physical_model_layer.h [hunk 3]
```diff
@@ -35,12 +39,83 @@ class PhysicalModelLayer : public ContainerLayer {
...
+
+  // |flow::PhysicalLayerShape|
+  const SkPath getPath() const;
```
Comment (chinmaygarde) [inline flow/layers/physical_model_layer.h]: This doesn't look right. Are you sure you didn't just mean `const SkPath &`?

File: flow/layers/physical_model_layer.h [hunk 4]
```diff
@@ -35,12 +39,83 @@ class PhysicalModelLayer : public ContainerLayer {
...
+#if defined(OS_FUCHSIA)
+  // |flow::PhysicalLayerShape|
+  const SkRRect& getFrameRRect() const { return rrect_; }
+#endif  // defined(OS_FUCHSIA)
+
+ private:
+  SkRRect rrect_;
+};
+
+class PhysicalLayerPath : public PhysicalLayerShape {
```
Comment (chinmaygarde) [inline flow/layers/physical_model_layer.h]: Nit: Can we make these subclasses final?

**Comment by chinmaygarde** [comment]
This seems a bit more complicated than I expected. We could make the physical model layer take an optional shape and perform the checks in manually in the paint and update scene. But I must admit, I am probably missing some background here.

File: flow/layers/physical_model_layer.h (hunk 2 above)
Comment (amirh) [reply flow/layers/physical_model_layer.h]: done

File: flow/layers/physical_model_layer.h (hunk 3 above)
Comment (amirh) [reply flow/layers/physical_model_layer.h]: The PhysicalLayerRRect implementation creates an SkPath on it's stack, so I cannot return a reference.
According to SkPath's docs copying should be cheap

Comment (chinmaygarde) [reply flow/layers/physical_model_layer.h]: In that case, the signature would be `SkPath getPath() const` right?

File: flow/layers/physical_model_layer.h (hunk 4 above)
Comment (amirh) [reply flow/layers/physical_model_layer.h]: done

**Comment by amirh** [comment]
Following f2f with Chinmay - keeping the PhysicalLayerShape classes as we must have specialized handling for path vs rrect cases in UpdateScene.

@chinmaygarde PTAL

File: flow/layers/physical_model_layer.h (hunk 3 above)
Comment (amirh) [reply flow/layers/physical_model_layer.h]: oops

**Pull request merged**